src/main/java/net/starschema/clouddb/jdbc/BQStatement.java

```diff
@@ -135,39 +135,32 @@ private ResultSet executeQueryHelper(String querySql, boolean unlimitedBillingBy
-                referencedJob = fetchedAll || this.connection.isClosed() ?
-                        null :
-                        this.connection.getBigquery()
-                                .jobs()
-                                .get(this.ProjectId, qr.getJobReference().getJobId())
-                                .setLocation(qr.getJobReference().getLocation())
-                                .execute();
-                if (qr.getJobComplete()) {
-                    if (resultSetType != ResultSet.TYPE_SCROLL_INSENSITIVE) {
-                        return new BQForwardOnlyResultSet(
-                                this.connection.getBigquery(),
-                                this.ProjectId.replace("__", ":").replace("_", "."),
-                                referencedJob, this, qr.getRows(), fetchedAll, qr.getSchema());
-                    } else if (fetchedAll) {
-                        // We can only return scrollable result sets here if we have all the rows: otherwise we'll
-                        // have to go get more below
-                        return new BQScrollableResultSet(qr.getRows(), this, qr.getSchema());
-                    }
-                    jobAlreadyCompleted = true;
+            QueryResponse qr = runSyncQuery(querySql, unlimitedBillingBytes);
+            boolean fetchedAll = qr.getJobComplete() && qr.getTotalRows() != null &&
+                    (qr.getTotalRows().equals(BigInteger.ZERO) ||
+                            (qr.getRows() != null && qr.getTotalRows().equals(BigInteger.valueOf(qr.getRows().size()))));
+            // Don't look up the job if we have nothing else we need to do
+            referencedJob = fetchedAll || this.connection.isClosed() ?
+                    null :
+                    this.connection.getBigquery()
+                            .jobs()
+                            .get(this.ProjectId, qr.getJobReference().getJobId())
+                            .setLocation(qr.getJobReference().getLocation())
+                            .execute();
+            if (qr.getJobComplete()) {
+                if (resultSetType != ResultSet.TYPE_SCROLL_INSENSITIVE) {
+                    return new BQForwardOnlyResultSet(
+                            this.connection.getBigquery(),
+                            this.ProjectId.replace("__", ":").replace("_", "."),
+                            referencedJob, this, qr.getRows(), fetchedAll, qr.getSchema());
+                } else if (fetchedAll) {
+                    // We can only return scrollable result sets here if we have all the rows: otherwise we'll
+                    // have to go get more below
+                    return new BQScrollableResultSet(qr.getRows(), this, qr.getSchema());
                 }
-            } else {
-                // Run the query async and return a Job that represents the running query
-                referencedJob = startQuery(querySql, unlimitedBillingBytes);
```

Comment:
Can you remove startQuery now?  Looks like it's used in CancelTest, but that could probably be changed to use the BQSupportFuncts version instead (which BQStatement.startQuery just calls out to).

Reply:
Removed